transport/transporters/rabbitmq/factory.go

```diff
@@ -0,0 +1,200 @@
+/*
+  Copyright 2019 Nextdoor.com, Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package rabbitmq
+
+import (
+	"net/url"
+	"os"
+
+	"github.com/NeowayLabs/wabbit"
+	"github.com/NeowayLabs/wabbit/amqp"
+	"github.com/Nextdoor/pg-bifrost.git/shutdown"
+	"github.com/Nextdoor/pg-bifrost.git/stats"
+	"github.com/Nextdoor/pg-bifrost.git/transport"
+	"github.com/Nextdoor/pg-bifrost.git/transport/batch"
+	"github.com/Nextdoor/pg-bifrost.git/transport/transporters/rabbitmq/transporter"
+	"github.com/cevaris/ordered_map"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	logger = logrus.New()
+	log    = logger.WithField("package", "rabbitmq")
+)
+
+type config struct {
+	host      string
+	username  string
+	password  string
+	vhost     string
+	exchange  string
+	batchSize int
+}
+
+func init() {
+	logger.SetOutput(os.Stdout)
+	logger.SetLevel(logrus.InfoLevel)
+}
+
+// New creates a rabbitmq transport which has rabbitmq batching and rabbitmq transporting, connected by channels
+func New(shutdownHandler shutdown.ShutdownHandler,
+	txnsWritten chan<- *ordered_map.OrderedMap, // Map of <transaction:progress.Written>
+	statsChan chan stats.Stat,
+	workers int,
+	inputChans []<-chan transport.Batch,
+	transportConfig map[string]interface{}) []*transport.Transporter {
+
+	conf := getConfig(transportConfig)
+
+	conn := createConnection(conf.host, conf.username, conf.password, conf.vhost, log)
```

Comment:
With the Connection-Channel model in RabbitMQ, how are network failures and reconnects handled on the TCP level in the `wabbit` library? And how are we handling those particular errors that propagate up? Since it looks like all of our Transport threads will be sharing a single connection.

Reply:
That is the normal model with RabbitMQ.  One connection and then each reader/writer opens a channel inside that connection.

The only time that is not suggested is if you want to avoid TCP pushback on publishing affecting the ability to consume messages.  That's not the case here where we are only publishing.  (https://derickbailey.com/2014/03/26/2-lessons-learned-and-3-resources-for-for-learning-rabbitmq-on-nodejs/)

Reply:
I see what you mean though.  I've added a shared ConnectionManager that listens for connection failures and will create a new connection for all the transports to create new channels from as needed.